### PR TITLE
Added PreviewLinkAutopopulate plugin to auto-populate preview links

### PR DESCRIPTION
**Pull request by stephen-cox**
Part of localgovdrupal/localgov#600

### CODE REVIEW

**Comment by ekes** [comment]
Is there something else that needs to be done other than checking out the branch. I've got it, and done a site install, and used demo, but I don't see the button, only one to copy the preview link.

Also is it intentionally only getting the next level of child pages, or should it be getting all child pages (using the Nested Storage)?

**Comment by stephen-cox** [comment]
> Also is it intentionally only getting the next level of child pages, or should it be getting all child pages (using the Nested Storage)?

I missed that subsite pages can also be parents along with overviews, I'll have to update to logic to handle this.

**Comment by finnlewis** [comment]
@stephen-cox working on this today

**Comment by stephen-cox** [comment]
@ekes I have updated this PR to use Entity Hierarchy to load the nodes in a subsite; which is frankly a pain to work with. Loading all nodes is not straightforward so I've added a test.

Would be nice to get this reviewed before Next Tuesday, if possible, as it's been hanging around for a while now.

File: src/Plugin/Block/SubsitesHierarchyTrait.php [hunk 1]
```diff
@@ -123,4 +123,41 @@ protected function getRootId(EntityInterface $entity): ?int {
...
+   *
+   * @return array
+   *   Flattened list of nodes in subsite hierarchy.
+   */
+  public function getFlattenedSubsiteHierarchy(NodeInterface $node): array {
+    $nodes = [];
+
+    $storage = $this->getNestedSetStorage('localgov_subsites');
+    $node_key = $this->getNestedSetNodeKeyFactory()->fromEntity($node);
+    if ($ancestors = $storage->findAncestors($node_key)) {
```
Comment (ekes) [inline src/Plugin/Block/SubsitesHierarchyTrait.php:143]: This could just be $storage->findRoot($node_key) which does much the same.

**Review by ekes — APPROVED** [review]
For the rest that works.

**Comment by ekes** [comment]
> Loading all nodes is not straightforward so I've added a test.

It's just get root, get all ancestors of root, and load them? Seems straightforward to me. Pain I found is the way that NestedSet class has be abstracted so you don't get code completion (or I don't anyway).

**Pull request merged**